### ISSUE TITLE
Consolidate on one default value for `fallbackToCacheTimeout`

### ISSUE THREAD

**Issue by brentvatne** [expo/expo#9368]
### Summary of Issue

Currently we default to `30000` if no option is provided in app.json. In templates, we provide `"fallbackToCacheTimeout": 0`. This means we effectively have two defaults.

In my experience, it is almost never correct to have your users wait up to 30 seconds on the splash screen for their app to load. It's very easy to get into a situation where this happens if you happen to intermittently have a bad internet connection.  Even in the best case, we need to check for an update before launching and this requires a roundtrip to Expo servers. On slow connections, this can mean adding seconds to startup, and in the best case maybe a couple hundred ms.

So, I think we should consolidate these defaults into one: always default to `fallbackToCacheTimeout` value of `0`.

## To make this change

- Update iOS / Android clients and/or expo-updates, where the default value lives iirc
- Change the schema default from 30000 to 0
- Add a step to `expo upgrade` from SDK <= 38 to add `updates.fallbackToCacheTimeout` `30000` if there is not one specified in app.json (if using app.config.js, log to the developer that they need to make this change manually).
- Update the [Configuring OTA Updates guide](https://docs.expo.io/guides/configuring-ota-updates/)

**Comment by shamilovtim** [comment]
I'd much rather risk a user getting stuck on the splash screen when their internet fails and odds are they will think there's something wrong with their phone than to risk not shipping users the very latest bundle of my application. With the new default setting being 0 I was completely puzzled by the latest bundle never appearing on time and it turned out that with this setting the newest bundle only appears after two restarts. In my opinion that completely defeats the purpose of expo-updates and as such doesn't quite make sense as a default settings. Either way this is a huge breaking change that I wish was in the migration guide from ExpoKit to Expo and was probably my main stumbling block (aside from the crash I'm having [here](https://forums.expo.io/t/after-migration-from-expokit-to-bare-the-new-build-crashes-on-testflight/40605/4) and still cannot debug which I'm starting to suspect is also related to expo-updates.

**Comment by brentvatne** [comment]
> I'd much rather risk a user getting stuck on the splash screen when their internet fails (and odds are they will think there's something wrong with their phone rather than blame the app) than to risk not shipping users the very latest bundle of my application

i think you are in the minority in this regard - this is very dependent on the application.

> With the new default setting being 0 I was very puzzled by the latest bundle never appearing on time and it turned out that with this setting the newest bundle only appears after two restarts.

interesting, we didn't actually change anything here. this is how it has been for years, i'm proposing defaulting to 0 in the future. right now we only do this in app.json in new projects.

> In my opinion that completely defeats the purpose of expo-updates and as such doesn't quite make sense as a default setting

let'd distinguish between low priority updates and high priority updates.

low priority updates should load in the background and become active whenever is convenient to do so. currently this means checking and downloading when the app starts, and in the future, checking in the background even when the app isn't running.

high priority updates often you want to stop the user from running the app at all until it can run. eg: you accidentally shipped a bad bug to production.

some apps always want to have high priority updates, eg: your app. but most apps, in my experience, end up publishing some low and some high priority updates (but without explicitly distinguishing between them because we don't have any expo-updates-native way to do that right now). most updates end up being low priority. some small fix to copy, an added feature, things like that. but sometimes you really need to get something in front of users right away. we're thinking about good ways to handle these very different update cases. right now users would need to take full control with manual updates, and associate some metadata with an update via some external api in order to treat it as high priority.

**Comment by jakobsen9** [comment]
Hi @brentvatne! 
I was looking for a reasonable value for the fallback when I got over this. 
Has there been any progress in easily distinguishing between low and high priority updates like you describe? As you say, we end up using the default of 0 for all updates, which are unfortunate when we accidentally push bugs, crashing our app. We want the app start time to be as fast as possible, so I'm quite reluctant to increasing the fallback timeout, making all updates high priority.

**Comment by brentvatne** [comment]
@jakobsen9 - you can use manual updates to do this: https://docs.expo.io/guides/configuring-ota-updates/#manual-updates

if there is an update available, you can get the manifest for the update in the `UpdateCheckResult` (`update` in `const update = await Updates.checkForUpdateAsync()`): https://github.com/expo/expo/blob/8e299d47cd867d36198e952bea16686aadb7d8da/packages/expo-updates/src/Updates.types.ts#L14

so you could attach some metadata like `highPriority: true` in your app config (possibly under the `extra` field) to do this

**Comment by markymc** [comment]
Just a note that the disparity between the documentation's mention of 30000ms default and the value that was put into the newly created app caused confusion for our team too. I definitely agree that having them both be either 30000ms or 0ms makes sense.

**Comment by brentvatne** [comment]
documentation updated, and we also default to 0 now, which is the same as in the template

**Comment by iway1** [comment]
> high priority updates often you want to stop the user from running the app at all until it can run. eg: you accidentally shipped a bad bug to production.

@brentvatne why not build this into EAS updates? It's very very common to want to distinguish between low and high priority and handle them differently. Expo could just offer:
```json
{
  fallbackToCacheTimeoutLowPriority: 3000,
  fallbackToCacheTimeoutHighPriority: 30000,
}
```

For example. Solves the problem. Eas CLI gets new option `eas update --priority high`. This would make our lives a lot easier. Right now EAS updates really require the dev to rack their brain a lot to get a good setup.

**Comment by brentvatne** [comment]
@iway1 - i personally believe that, unless the app is stuck in a crash loop, blocking booting the app on waiting for a network request to check for an update is almost never desirable. the app developer should provide a better experience for downloading a critical update than delaying every single app launch to check if a low priority or a critical update exists. you can build this and have full control over the experience using the apis that exist today. this example app demonstrates one approach to doing so: https://github.com/expo/UpdatesAPIDemo?tab=readme-ov-file#push-a-critical-update

**Comment by iway1** [comment]
@brentvatne I think there is a lot to think about this. Some thoughts

- Showing a "Please update app" popup that distracts the user and then requires bundle restart is IMO a much worse user experience than delaying app boot by ~2 seconds or whatever it takes, and then having the user use the app normally.

- Why does *every* app launch need to be delayed significantly? Is there no fast way to check that an update is available that happens consistently in <100ms? if that were possible app startup times would only be delayed meaningfully when necessary. 

- Users almost never kill an app entirely, so `fallbackToCacheTimeout` is *mostly* only relevant for new app installs / app updates... in our case the entire reason we want to use it is to prevent a "please update" popup from happening on a fresh install... so it's really about improving the UX for us. 

It's really hard to get a great UX for updates without having them happen in the background or happen really fast. I'm also just confused on why download times take so long for our app... like the exported bundle with assets for us is 10MB, is EAS Updates redownloading all assets in the bundle every time an update is shipped?

**Comment by iway1** [comment]
anyways I really wish the expo docs told me how this stuff actually worked it's hard to make informed decisions about which behavior to use without knowing how it works under he hood

**Comment by brentvatne** [comment]
> Why does every app launch need to be delayed significantly? Is there no fast way to check that an update is available that happens consistently in <100ms?

no, that's not possible. sure, you could limit it to 100ms but then it would not be very useful, only users on very fast connections would be able to check for the update and download the manifest. on slow connections, or lie-fi (eg: bad wifi with lots of packet loss in hotels/busses/trains/etc) you would always exceed the 100ms.

> Users almost never kill an app entirely, so fallbackToCacheTimeout is mostly only relevant for new app installs / app updates... in our case the entire reason we want to use it is to prevent a "please update" popup from happening on a fresh install... so it's really about improving the UX for us.

is the ux really better if you're stuck on the splash screen for 30 seconds, with no indication of any sort of progress, just because there is a high priority update?

**Comment by brentvatne** [comment]
> anyways I really wish the expo docs told me how this stuff actually worked it's hard to make informed decisions about which behavior to use without knowing how it works under he hood

what do you want to know about specifically? we can certainly add more documentation! i think explicitly calling out this case of building high priority updates and some ux for it would be a great add

**Comment by iway1** [comment]
>is the ux really better if you're stuck on the splash screen for 30 seconds, with no indication of any sort of progress, just because there is a high priority update?

In the 30 second case, no it is not, but I think there is some speed at which it is better. Totally agree 30 seconds is too long in any case.

> what do you want to know about specifically? we can certainly add more documentation! i think explicitly calling out this case of building high priority updates and some ux for it would be a great add

@brentvatne the thing I am mainly interested in is that I have no idea how to figure out, how do we reduce the time it takes to download an update? What factors make an update take significantly longer? Does EAS have consistent update download speeds or does server load make it some times take a lot longer?

For example I know that:

- Updates include assets
- There is an update manifest

What I don't know is:

- Does an update require the front end to download *all* assets included in the update (which is all assets in the project), or does the front end only download the assets it does not already have? This would mean we should really optimize our asset sizes as a priority.
- What requests does the front end make when downloading an update? Is it: `download manifest -> download update if necessary as determined by manifest`?

And importantly:

- What is the recommended way of implementing updates?

I feel like the community would really benefit from some guidance in this respect. When I was building out our eas updates implementation I felt very in the dark about what the best way to implement it would be for our users and business, so I just had to sort of come up with out implementation from scratch.

I feel like by default, most devs don't even really think about how OTA updates affect their users and in what cases they can expect the latest code to be running on their application. There is actually a lot to think about to decide what's best for each app.

**Comment by wschurman** [comment]
Hi @iway1 - thanks for the feedback!

Most of your questions I think are answered by https://docs.expo.dev/eas-update/how-it-works/#practical-overview but I'll also answer them here.

> how do we reduce the time it takes to download an update? What factors make an update take significantly longer? Does EAS have consistent update download speeds or does server load make it some times take a lot longer?

The main factor in time is download size. The expo-updates library and EAS Update do their best to compress and parallelize requests but in general the fewer assets that are updated in an OTA update the faster downloading and applying a new will be.

> Does an update require the front end to download all assets included in the update (which is all assets in the project), or does the front end only download the assets it does not already have? This would mean we should really optimize our asset sizes as a priority.

It downloads all updates not previously downloaded or in the original build. By default, at build time, all assets are embedded in the build, so releasing code-only updates (no new images/etc) will only download the manifest and the new JS bundle.

> What requests does the front end make when downloading an update? Is it: download manifest -> download update if necessary as determined by manifest?

Correct!

> Does .fallbackToCacheTimeout take effect when an app is updated from an App Store rather than a fresh install?

Not sure I follow. Let me know if https://docs.expo.dev/eas-update/how-it-works/#practical-overview doesn't answer this.

> What is the recommended way of implementing updates?

This is definitely an area we're trying to figure out what a good default and workflow are for updates/releases/previews/etc. The main challenge is that it's fairly tied to how a development team does things like code review, branch cuts, app versioning, dogfooding, QA, and releasing, so it's hard to prescribe a single recommended way of implementing updates. That being said, we need do better to onboard developers who desire us to prescribe a one-size-fits-most solution out of the box, and that's something we're actively thinking about. There's the https://docs.expo.dev/eas-update/deployment-patterns/ but even that is a bit low-level, whereas maybe we need a more sensible default.

**Comment by wschurman** [comment]
As far as update priority, it's definitely something we've considered. It's difficult to build a solution that fits all uses cases, though. For example, here's some examples of use cases we've heard:
- An app that wants to prioritize their network requests (the network requests for their main screen for example) over expo-updates downloading except when an update is high priority.
- An app that wants to prevent hiding the splash screen until update is downloaded when update is high priority.
- An app that wants to delay downloading until after main screen has loaded, and then download the update, and then present a dialog to the user telling them to restart if the update is high priority.
- An app that wants to delay downloading until after main screen has loaded, and then if the downloaded update is high priority call `Updates.reload` to force a "relaunch".
- ... (lots more contrived examples)

What we determined is that while some cases may be generalizable to a concept like `fallbackToCacheTimeoutLowPriority`/`fallbackToCacheTimeoutHighPriority`, unfortunately not all cases can, and in fact most cannot. So what we decided to do for now was document one example of how it may be implemented in accordance with one's product requirements (https://github.com/expo/UpdatesAPIDemo?tab=readme-ov-file#push-a-critical-update). But it's a decision that we re-discuss fairly often so it's good to know that there's a desire for a native feature of this sort that does a basic prioritization.
